lib/fs/types.go

```diff
@@ -6,20 +6,37 @@
 
 package fs
 
-type FilesystemType int32
+import "sync"
 
-const (
-	FilesystemTypeBasic FilesystemType = 0
-	FilesystemTypeFake  FilesystemType = 1
-)
+type FilesystemType string
 
-func (t FilesystemType) String() string {
-	switch t {
-	case FilesystemTypeBasic:
-		return "basic"
-	case FilesystemTypeFake:
-		return "fake"
-	default:
-		return "unknown"
-	}
+// Option modifies a filesystem at creation. An option might be specific
+// to a filesystem-type.
+//
+// String is used to detect options with the same effect, i.e. must be different
+// for options with different effects. Meaning if an option has parameters, a
+// representation of those must be part of the returned string.
+type Option interface {
+	String() string
+	apply(Filesystem) Filesystem
+}
+
+// Factory function type for constructing a custom file system. It takes the following
+// parameters:
+// 1. The fsType (string, e.g. "basic", "fake")
+// 2. The file system URI (string)
+// 3. Options for the file system
+type FilesystemFactory func(FilesystemType, string, ...Option) (Filesystem, error)
```

Comment:
I don't think the factory needs to get the FilesystemType as a parameter, as it is tied to the type we registered it under, slightly simplifying the interface. (If we had the same factory function for lots of different types for some reason we could wrap it when registering instead.)